### ISSUE TITLE
Vlen strings in attributes are not decoded after applying NCO

### ISSUE THREAD

**Issue by dionhaefner** [h5netcdf/h5netcdf#116]
Example script:

```python
import h5netcdf
import netCDF4
import subprocess

with h5netcdf.File("test.nc", "w", decode_vlen_strings=True) as f:
    f.attrs["foo"] = "bar"

subprocess.run(["ncks", "test.nc", "-o", "test2.nc", "--overwrite"])

with h5netcdf.File("test.nc", "r", decode_vlen_strings=True) as f:
    print("attrs original file:")
    print(f.attrs)

with h5netcdf.File("test2.nc", "r", decode_vlen_strings=True) as f:
    print("\nattrs changed file:")
    print(f.attrs)

with netCDF4.Dataset("test2.nc", "r") as f:
    print("\nattrs changed file with netCDF4:")
    for attr in f.ncattrs():
        print(f"{attr}: {getattr(f, attr)!r}")

```

Output:

```bash
$ python h5netcdf_bug.py
attrs original file:
<class 'h5netcdf.attrs.Attributes'>
foo: 'bar'

attrs changed file:
<class 'h5netcdf.attrs.Attributes'>
foo: array(['bar'], dtype=object)
history: b'Tue Nov  2 11:57:35 2021: ncks test.nc -o test2.nc --overwrite'
NCO: b'netCDF Operators version 5.0.3 (Homepage = http://nco.sf.net, Code = http://github.com/nco/nco)'

attrs changed file with netCDF4:
foo: 'bar'
history: 'Tue Nov  2 11:57:35 2021: ncks test.nc -o test2.nc --overwrite'
NCO: 'netCDF Operators version 5.0.3 (Homepage = http://nco.sf.net, Code = http://github.com/nco/nco)'
```

In the changed file, `h5netcdf` doesn't decode the `foo` attribute anymore (`array(['bar'], dtype=object)`). NetCDF4 decodes everything.

NetCDF files: [ncfiles.zip](https://github.com/h5netcdf/h5netcdf/files/7460017/ncfiles.zip)

This is with most recent h5netcdf and h5py.

**Comment by kmuehlbauer** [comment]
There is a difference in the underlying hdf-representation between those two files. The first has `SCALAR` DATASPACE whereas the second has `SIMPLE` (array). That's why an array is returned.

test.nc
```
ATTRIBUTE "foo" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "bar"
      }
   }
```

test2.nc

```
ATTRIBUTE "foo" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
      DATA {
      (0): "bar"
      }
   }
```

One additional note: `decode_vlen_strings` has only effect when extracting Datasets (means Variables, not Attributes). For attributes this should give the same output regardless the value of `decode_vlen_strings`.

Technically we just rely on `h5py` to correctly decode attribute strings:

> Variable-length strings in attributes are read as str objects. These are decoded as UTF-8 with surrogate escaping for unrecognised bytes.

**Comment by dionhaefner** [comment]
I see. It's a bit of a bummer since the modified file loads just fine with netCDF4, so you get counter-intuitive behavior downstream:

```python
import xarray as xr

ds = xr.open_dataset("test.nc", engine="netcdf4")
ds.var.plot()  # works

ds = xr.open_dataset("test.nc", engine="h5netcdf")
ds.var.plot()  # throws an exception because units can't be parsed
```

**Comment by kmuehlbauer** [comment]
A possible solution on the `h5netcdf` side would be to check if the array is of length 1 subset that one element. But I'm not really sure, if this would be the correct place to handle this issue.

**Comment by dionhaefner** [comment]
That seems to be how netcdf-python does it:

https://github.com/Unidata/netcdf4-python/blob/5ed431ceb0e594bccabe625e8782381e79b284de/src/netCDF4/_netCDF4.pyx#L1503-L1508

**Comment by kmuehlbauer** [comment]
I'm wondering if we should not put this under the `decode_vlen_strings` umbrella and do the decoding depending on the keyword? @shoyer Thoughts?

**Comment by kmuehlbauer** [comment]
I've tampered with this a bit more. Here are two different issues, which at first I did not pull apart correctly.

`array(['bar'], dtype=object)` does not mean it is encoded (representing some `bytes`) but just a string array. The other thing is that the other attributes (history/NCO) are really not decoded.

I still think this solvable.

**Comment by kmuehlbauer** [comment]
Just to not forget about this:

The following snippet writes either netcdf4-style NULLTERM or h5py-style NULLPAD fixed sized string attributes (only which translates to ascii). So if we apply this (eg. at the `legacyapi`), we could mimic netcdf4 fixed size attributes. We would need to correctly decode when reading.

- NULLTERM - `term=h5py.h5t.STR_NULLTERM` (default in this case)
- NULLPAD - `term=h5py.h5t.STR_NULLPAD`

```python
def write_string_attr(loc, name, value, term=None):
    tid = h5py.h5t.C_S1.copy()
    tid.set_size(len(value))
    if term is not None:
        tid.set_strpad(term)
    sid = h5py.h5s.create(h5py.h5s.SCALAR)
    value = np.array(value, dtype="|S")
    aid = h5py.h5a.create(loc._h5group.id, name.encode(), tid, sid)
    aid.write(value, mtype=aid.get_type())
```

The other problem with the array of strings is that netcdf4-python is creating VLEN String Attributes which have a SIMPLE dataspace for encoded/bytes string. Normal Python strings have SCALAR dataspace. `h5py` creates UTF8 from encoded/bytes as well as normal Python strings with SCALAR dataspace. To create SIMPLE dataspace one would need to provide an array of strings.

@dionhaefner If you are up to it, we can discuss the needed changes in h5netcdf to implement it in a decent way.

Example code:

```python
import h5py

def write_string_attr(loc, name, value, term=None):
    tid = h5py.h5t.C_S1.copy()
    tid.set_size(len(value))
    if term is not None:
        tid.set_strpad(term)
    sid = h5py.h5s.create(h5py.h5s.SCALAR)
    value = np.array(value, dtype="|S")
    aid = h5py.h5a.create(loc._h5group.id, name.encode(), tid, sid)
    aid.write(value, mtype=aid.get_type())

with legacyapi.Dataset("test1.nc", "w") as f:
    f.unicode = u"foé"
    f.bytes = b"bytes"
    f.string = "baz"
    f.array = ["arr"]
    write_string_attr(f, "nc4_char", "foo", term=h5py.h5t.STR_NULLTERM)
    write_string_attr(f, "h5_char", "foo", term=h5py.h5t.STR_NULLPAD)

subprocess.run(["ncks", "test1.nc", "-o", "test2.nc", "--overwrite"])

!h5dump test1.nc
!h5dump test2.nc
!ncdump test1.nc
!ncdump test2.nc


with h5netcdf.File("test1.nc", "r", decode_vlen_strings=True) as f:
    print("\nattrs original file with h5netcdf:")
    print(f.attrs)
    
with legacyapi.Dataset("test1.nc", "r") as f:
    print("\nattrs original file with legacyapi:")
    print(f.attrs)

with netCDF4.Dataset("test1.nc", "r") as f:
    print("\nattrs original file with netCDF4:")
    print(f)
    for attr in f.ncattrs():
        print(f"{attr}: {getattr(f, attr)!r}")

with h5netcdf.File("test2.nc", "r", decode_vlen_strings=True) as f:
    print("\nattrs changed file with h5netcdf:")
    print(f.attrs)
    
with legacyapi.Dataset("test2.nc", "r") as f:
    print("\nattrs changed file with legacyapi:")
    print(f.attrs)

with netCDF4.Dataset("test2.nc", "r") as f:
    print("\nattrs changed file with netCDF4:")
    for attr in f.ncattrs():
        print(f"{attr}: {getattr(f, attr)!r}")
```

Output:

```
HDF5 "test1.nc" {
GROUP "/" {
   ATTRIBUTE "_NCProperties" {
      DATATYPE  H5T_STRING {
         STRSIZE 48;
         STRPAD H5T_STR_NULLPAD;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "version=2,h5netcdf=0.13.0,hdf5=1.12.1,h5py=3.4.0"
      }
   }
   ATTRIBUTE "array" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
      DATA {
      (0): "arr"
      }
   }
   ATTRIBUTE "bytes" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "bytes"
      }
   }
   ATTRIBUTE "h5_char" {
      DATATYPE  H5T_STRING {
         STRSIZE 3;
         STRPAD H5T_STR_NULLPAD;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "foo"
      }
   }
   ATTRIBUTE "nc4_char" {
      DATATYPE  H5T_STRING {
         STRSIZE 3;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "foo"
      }
   }
   ATTRIBUTE "string" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "baz"
      }
   }
   ATTRIBUTE "unicode" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "fo\37777777703\37777777651"
      }
   }
}
}
HDF5 "test2.nc" {
GROUP "/" {
   ATTRIBUTE "NCO" {
      DATATYPE  H5T_STRING {
         STRSIZE 96;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "netCDF Operators version 5.0.5 (Homepage = http://nco.sf.net, Code = http://github.com/nco/nco)"
      }
   }
   ATTRIBUTE "_NCProperties" {
      DATATYPE  H5T_STRING {
         STRSIZE 34;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "version=2,netcdf=4.8.1,hdf5=1.12.1"
      }
   }
   ATTRIBUTE "array" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
      DATA {
      (0): "arr"
      }
   }
   ATTRIBUTE "bytes" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
      DATA {
      (0): "bytes"
      }
   }
   ATTRIBUTE "h5_char" {
      DATATYPE  H5T_STRING {
         STRSIZE 3;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "foo"
      }
   }
   ATTRIBUTE "history" {
      DATATYPE  H5T_STRING {
         STRSIZE 64;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "Tue Feb  1 15:17:46 2022: ncks test1.nc -o test2.nc --overwrite"
      }
   }
   ATTRIBUTE "nc4_char" {
      DATATYPE  H5T_STRING {
         STRSIZE 3;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "foo"
      }
   }
   ATTRIBUTE "string" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
      DATA {
      (0): "baz"
      }
   }
   ATTRIBUTE "unicode" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SIMPLE { ( 1 ) / ( 1 ) }
      DATA {
      (0): "fo\37777777703\37777777651"
      }
   }
}
}
netcdf test1 {

// global attributes:
		string :unicode = "foé" ;
		string :bytes = "bytes" ;
		string :string = "baz" ;
		string :array = "arr" ;
		:nc4_char = "foo" ;
		:h5_char = "foo" ;
}
netcdf test2 {

// global attributes:
		string :unicode = "foé" ;
		string :bytes = "bytes" ;
		string :string = "baz" ;
		string :array = "arr" ;
		:nc4_char = "foo" ;
		:h5_char = "foo" ;
		:history = "Tue Feb  1 15:17:46 2022: ncks test1.nc -o test2.nc --overwrite" ;
		:NCO = "netCDF Operators version 5.0.5 (Homepage = http://nco.sf.net, Code = http://github.com/nco/nco)" ;
}

attrs original file with h5netcdf:
<class 'h5netcdf.attrs.Attributes'>
array: array(['arr'], dtype=object)
bytes: 'bytes'
h5_char: b'foo'
nc4_char: b'foo'
string: 'baz'
unicode: 'foé'

attrs original file with legacyapi:
<class 'h5netcdf.attrs.Attributes'>
array: array(['arr'], dtype=object)
bytes: 'bytes'
h5_char: b'foo'
nc4_char: b'foo'
string: 'baz'
unicode: 'foé'

attrs original file with netCDF4:
<class 'netCDF4._netCDF4.Dataset'>
root group (NETCDF4 data model, file format HDF5):
    unicode: foé
    bytes: bytes
    string: baz
    array: arr
    nc4_char: foo
    h5_char: foo
    dimensions(sizes): 
    variables(dimensions): 
    groups: 
unicode: 'foé'
bytes: 'bytes'
string: 'baz'
array: 'arr'
nc4_char: 'foo'
h5_char: 'foo'

attrs changed file with h5netcdf:
<class 'h5netcdf.attrs.Attributes'>
unicode: array(['foé'], dtype=object)
bytes: array(['bytes'], dtype=object)
string: array(['baz'], dtype=object)
array: array(['arr'], dtype=object)
nc4_char: b'foo'
h5_char: b'foo'
history: b'Tue Feb  1 15:17:46 2022: ncks test1.nc -o test2.nc --overwrite'
NCO: b'netCDF Operators version 5.0.5 (Homepage = http://nco.sf.net, Code = http://github.com/nco/nco)'

attrs changed file with legacyapi:
<class 'h5netcdf.attrs.Attributes'>
unicode: array(['foé'], dtype=object)
bytes: array(['bytes'], dtype=object)
string: array(['baz'], dtype=object)
array: array(['arr'], dtype=object)
nc4_char: b'foo'
h5_char: b'foo'
history: b'Tue Feb  1 15:17:46 2022: ncks test1.nc -o test2.nc --overwrite'
NCO: b'netCDF Operators version 5.0.5 (Homepage = http://nco.sf.net, Code = http://github.com/nco/nco)'

attrs changed file with netCDF4:
unicode: 'foé'
bytes: 'bytes'
string: 'baz'
array: 'arr'
nc4_char: 'foo'
h5_char: 'foo'
history: 'Tue Feb  1 15:17:46 2022: ncks test1.nc -o test2.nc --overwrite'
NCO: 'netCDF Operators version 5.0.5 (Homepage = http://nco.sf.net, Code = http://github.com/nco/nco)'
```

**Comment by dionhaefner** [comment]
What a hot mess. Sorry for dragging you into that 😄 

IIUC there are 2 separate issues: inconsistent writing, and inconsistent decoding. I think the writing one depends on the NetCDF4 spec: does it specify whether string attributes should have a `SIMPLE` dataspace, or is that an implementation detail? In the latter case I think it would be OK to ignore the writing inconsistency (as long as the outputs are valid NetCDF4 files). 

Consistent reading sounds more important in my opinion. Doesn't this mostly boil down to converting SIMPLE dataspace strings to Python str on read?

**Comment by kmuehlbauer** [comment]
> I think the writing one depends on the NetCDF4 spec: does it specify whether string attributes should have a `SIMPLE` dataspace, or is that an implementation detail?

Given that comment from @DennisHeimbigner https://github.com/Unidata/netcdf-c/issues/127#issuecomment-292625916, netcdf4 treats attributes as vectors of values, even if only one element. That means they should have SIMPLE dataspace.

> Consistent reading sounds more important in my opinion. Doesn't this mostly boil down to converting SIMPLE dataspace strings to Python str on read?

Yes, I think that assupmption is correct.

**Comment by kmuehlbauer** [comment]
Resolved by #151 for the reading part. Writing attributes with SIMPLE dataspace vs. SCALAR dataspace is not yet decided, as `netcdf-c`/`netcdf4-python` can read it out of the box.
